src/driver-gskill.c

```diff
@@ -0,0 +1,1365 @@
+	uint8_t brightness;
+	struct gskill_led_color dpi[4];
+	struct gskill_led_color leds[GSKILL_LED_TYPE_COUNT];
+} __attribute__((packed));
+
+struct gskill_dpi_led_group_cfg {
+	uint8_t duration_step;
+	uint8_t duration_high;
+	uint8_t duration_low;
+	uint8_t cycle_num;
+
+	struct gskill_led_values steps[12];
+};
+
+struct gskill_led_group_cfg {
+	enum gskill_led_control_type type :3;
+	uint8_t                           :5; /* unused */
+
+	uint8_t duration_step;
+	uint8_t duration_high;
+	uint8_t duration_low;
+	uint8_t cycle_num;
+
+	struct gskill_led_values steps[12];
+} __attribute__((packed));
+
+struct gskill_dpi_led_cycle_cfg {
+	enum gskill_led_control_type type :3;
+	uint8_t                           :5; /* unused */
+
+	/* Don't worry, the low/high flip-flop here is intentional */
+	uint8_t duration_low;
+	uint8_t duration_high;
+	uint8_t cycle_num;
+
+	struct gskill_led_values cycles[12];
+} __attribute__((packed));
+
+/*
+ * We may occasionally run into codes outside this, however those codes
+ * indicate functionalities that aren't too useful for us
+ */
+enum gskill_button_function_type {
+	GSKILL_BUTTON_FUNCTION_WHEEL                = 0x00,
+	GSKILL_BUTTON_FUNCTION_MOUSE                = 0x01,
+	GSKILL_BUTTON_FUNCTION_KBD                  = 0x02,
+	GSKILL_BUTTON_FUNCTION_CONSUMER             = 0x03,
+	GSKILL_BUTTON_FUNCTION_MACRO                = 0x06,
+	GSKILL_BUTTON_FUNCTION_DPI_UP               = 0x09,
+	GSKILL_BUTTON_FUNCTION_DPI_DOWN             = 0x0a,
+	GSKILL_BUTTON_FUNCTION_CYCLE_DPI_UP         = 0x0b,
+	GSKILL_BUTTON_FUNCTION_CYCLE_DPI_DOWN       = 0x0c,
+	GSKILL_BUTTON_FUNCTION_PROFILE_SWITCH       = 0x0d,
+	GSKILL_BUTTON_FUNCTION_TEMPORARY_CPI_ADJUST = 0x15,
+	GSKILL_BUTTON_FUNCTION_DIRECT_DPI_CHANGE    = 0x16,
+	GSKILL_BUTTON_FUNCTION_CYCLE_PROFILE_UP     = 0x18,
+	GSKILL_BUTTON_FUNCTION_CYCLE_PROFILE_DOWN   = 0x19,
+	GSKILL_BUTTON_FUNCTION_DISABLE              = 0xff
+};
+
+struct gskill_button_cfg {
+	enum gskill_button_function_type type :8;
+	union {
+		struct {
+			enum {
+				GSKILL_WHEEL_SCROLL_UP = 0,
+				GSKILL_WHEEL_SCROLL_DOWN = 1,
+			} direction :8;
+		} wheel;
+
+		struct {
+			enum {
+				GSKILL_BTN_MASK_LEFT   = (1 << 0),
+				GSKILL_BTN_MASK_RIGHT  = (1 << 1),
+				GSKILL_BTN_MASK_MIDDLE = (1 << 2),
+				GSKILL_BTN_MASK_SIDE   = (1 << 3),
+				GSKILL_BTN_MASK_EXTRA  = (1 << 4)
+			} button_mask :8;
+		} mouse;
+
+		struct {
+			uint16_t code;
+		} consumer;
+
+		struct {
+			uint8_t modifier_mask;
+			uint8_t hid_code;
+			/*
+			 * XXX: Supposedly this is supposed to have additional
+			 * parts of the kbd code, however that doesn't seem to
+			 * be the case in practice…
+			 */
+			uint16_t :16;
+		} kbd;
+
+		struct {
+			uint8_t level;
+		} dpi;
+	} params;
+} __attribute__((packed));
+
+struct gskill_action_mapping {
+	struct gskill_button_cfg config;
+	struct ratbag_button_action action;
+};
+
+struct gskill_profile_report {
+	uint16_t                  :16;
+	uint8_t profile_num;
```

Comment:
Gonna reply here since it looks like you've gone to bed. I can't actually find where the spec sheet lists the profile number, but looking at the raw profile data it's definitely the third byte on reads:
```
meerkat%:~/Projects/libratbag> sudo -E ./tools/ratbag-command --verbose=raw info /dev/input/event17
ratbag debug: G.SKILL Corporation G.SKILL RIPJAWS MX780 GAMING MOUSE is device '/dev/hidraw1'.
ratbag raw: feature set:   0c c4 12 00 01 00 00 00 00
ratbag raw: feature get:   00 b0 01 05 00 00 00 00 00
ratbag raw: feature set:   0c c4 08 00 00 00 00 00 00
ratbag raw: feature get:   00 b0 55 aa 15 00 00 00 00
ratbag debug: Firmware version: 21
ratbag raw: feature set:   0c c4 0c 00 00 00 00 00 00
ratbag raw: feature get:   b0 05 00 27 60 07 03 21 0c 0c 23 23 00 00 00 00 00 00 ff c6 01 0b c6 01 0b c6 01 0b c6 01 0b c6 01 0b c6 01 0b c6 01 0b 04 05 00 03 ff ff 00 00 ff 00 ff 00 ff 00 00 ff ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b 02 01 01 03 ff ff 00 00 ff 00 ff 00 ff 00 00 ff ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b 02 02 01 01 03 ff ff 00 00 ff 00 ff 00 ff 00 00 ff ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b 02 02 01 01 03 ff ff 00 00 ff 00 ff 00 ff 00 00 ff ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b 02 02 01 01 03 ff ff 00 00 ff 00 ff 00 ff 00 00 ff ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b 00 01 02 03 04 05 06 07 01 02 00 00 00 01 01 00 00 00 01 04 00 00 00 01 10 00 00 00 01 08 00 00 00 0b 00 00 00 00 01 10 00 00 00 01 08 00 00 00 00 00 00 00 00 00 01 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff
ratbag debug: Profile 0: DPI count is 2
ratbag raw: feature set:   0c c4 0c 01 00 00 00 00 00
ratbag raw: feature get:   b0 05 01 84 60 10 02 50 10 10 20 20 30 30 60 60 a4 a4 ff c6 01 0b c6 01 0b c6 01 0b c6 01 0b c6 01 0b c6 01 0b c6 01 0b 04 05 00 03 ff ff 00 00 ff 00 ff 00 ff 00 00 ff ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b 02 01 01 03 ff ff 00 00 ff 00 ff 00 ff 00 00 ff ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b 02 02 01 01 03 ff ff 00 00 ff 00 ff 00 ff 00 00 ff ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b 02 02 01 01 03 ff ff 00 00 ff 00 ff 00 ff 00 00 ff ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b 02 02 01 01 03 ff ff 00 00 ff 00 ff 00 ff 00 00 ff ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b 00 01 02 03 04 05 06 07 01 01 00 00 00 01 02 00 00 00 01 04 00 00 00 01 08 00 00 00 01 10 00 00 00 0b 00 00 00 00 ff 00 00 00 00 ff 00 00 00 00 00 00 00 00 00 00 01 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 48 00 65 00 79 00 54 00 68 00 65 00 72 00 65 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag debug: Profile 1: DPI count is 5
ratbag raw: feature set:   0c c4 0c 02 00 00 00 00 00
ratbag raw: feature get:   b0 05 02 a5 60 07 03 51 10 10 20 20 30 30 60 60 a4 a4 ff c6 01 0b c6 01 0b c6 01 0b c6 01 0b c6 01 0b c6 01 0b c6 01 0b 04 05 00 03 ff ff 00 00 ff 00 ff 00 ff 00 00 ff ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b 02 01 01 03 ff ff 00 00 ff 00 ff 00 ff 00 00 ff ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b 02 02 01 01 03 ff ff 00 00 ff 00 ff 00 ff 00 00 ff ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b 02 02 01 01 03 ff ff 00 00 ff 00 ff 00 ff 00 00 ff ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b 02 02 01 01 03 ff ff 00 00 ff 00 ff 00 ff 00 00 ff ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b 00 01 02 03 04 05 06 07 01 01 00 00 00 01 02 00 00 00 01 04 00 00 00 01 08 00 00 00 01 10 00 00 00 0b 00 00 00 00 ff 00 00 00 00 ff 00 00 00 00 00 00 00 00 00 00 01 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 50 00 72 00 6f 00 66 00 69 00 6c 00 65 00 33 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag debug: Profile 2: DPI count is 5
ratbag raw: feature set:   0c c4 0c 03 00 00 00 00 00
ratbag raw: feature get:   b0 05 03 a4 60 07 03 51 10 10 20 20 30 30 60 60 a4 a4 ff c6 01 0b c6 01 0b c6 01 0b c6 01 0b c6 01 0b c6 01 0b c6 01 0b 04 05 00 03 ff ff 00 00 ff 00 ff 00 ff 00 00 ff ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b 02 01 01 03 ff ff 00 00 ff 00 ff 00 ff 00 00 ff ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b 02 02 01 01 03 ff ff 00 00 ff 00 ff 00 ff 00 00 ff ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b 02 02 01 01 03 ff ff 00 00 ff 00 ff 00 ff 00 00 ff ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b 02 02 01 01 03 ff ff 00 00 ff 00 ff 00 ff 00 00 ff ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b 00 01 02 03 04 05 06 07 01 01 00 00 00 01 02 00 00 00 01 04 00 00 00 01 08 00 00 00 01 10 00 00 00 0b 00 00 00 00 ff 00 00 00 00 ff 00 00 00 00 00 00 00 00 00 00 01 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 50 00 72 00 6f 00 66 00 69 00 6c 00 65 00 34 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag debug: Profile 3: DPI count is 5
ratbag raw: feature set:   0c c4 0c 04 00 00 00 00 00
ratbag raw: feature get:   b0 05 04 71 60 10 02 51 10 10 20 20 30 30 60 60 a4 a4 ff c6 01 0b c6 01 0b c6 01 0b c6 01 0b c6 01 0b c6 01 0b c6 01 0b 04 05 00 03 ff ff 00 00 ff 00 ff 00 ff 00 00 ff ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b 02 01 01 03 ff ff 00 00 ff 00 ff 00 ff 00 00 ff ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b 02 02 01 01 03 ff ff 00 00 ff 00 ff 00 ff 00 00 ff ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b 02 02 01 01 03 ff ff 00 00 ff 00 ff 00 ff 00 00 ff ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b 02 02 01 01 03 ff ff 00 00 ff 00 ff 00 ff 00 00 ff ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b ff c6 01 0b 00 01 02 03 04 05 06 07 01 02 00 00 00 01 01 00 00 00 01 04 00 00 00 01 02 00 00 00 01 40 00 00 00 0b 00 00 00 00 ff 00 00 00 00 ff 00 00 00 00 00 00 00 00 00 00 01 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 50 00 72 00 6f 00 66 00 69 00 6c 00 65 00 35 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag debug: Profile 4: DPI count is 5
ratbag raw: feature set:   0c c4 07 00 01 00 00 00 00
ratbag raw: feature get:   00 b0 01 00 00 00 00 00 00
ratbag debug: driver match found: G.Skill Ripjaws MX780
Device 'G.SKILL Corporation G.SKILL RIPJAWS MX780 GAMING MOUSE'
```